doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,443 @@
+
+The proposal is to preserve this behaviour for the most part.  This is achieved
+by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
+in these operations.
+To mitigate user surprises, we would further make conversion to the new type
+more strict.  This means that the results will be unchanged in the first
+two examples.  For the second one, the results will be the following::
+
+  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
+  nparray([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity
+
+The second one will warn because ``np.float32(1e200)`` overflows to infinity.
+It will then do the calculation with ``inf`` as normally.
+
+
+Impact on functions involving Python ``int``, ``float``, and ``complex``
+------------------------------------------------------------------------
+
+Most functions, in particular ``ufuncs`` will also use this weakly typed
+logic.
+In some cases, functions will call `np.asarray()` on inputs before any operations
+and thus will s
+
+.. note::
+
+    There is a real alternative to not do this for `ufuncs` and limit the special
+    behaviour to Python operators.  From a user perspective, we assume that most
+    functions effectively call `np.asarray()`.
+    Because Python operators allow more custom logic, this would ensure that an
+    overflow warning is given for all results with decreased precision.
+
+
+Backward compatibility
+======================
+
+In general, code which only uses the default dtypes float64, or int32/int64
+or more precise ones should not be affected.
+
+However, the proposed changes will modify results in quite a few cases where
+0-D or scalar values (with non-default dtypes) are mixed.
+In many cases, these will be bug-fixes, however, there are certain changes
+which may be particularly interesting.
+
+The most important failure is probably the following example::
+
+  arr = np.arange(100, dtype=np.uint8)  # storage array with low precision
+  value = arr[10]
+
+  # calculation continues with "value" without considering where it came from
+  value * 100
+
+Where previously the ``value * 100`` would cause an up-cast to int32/int64
+(because value is a scalar).  The new behaviour will preserve the lower
+precision unless explicitly dealt with (just as if ``value`` was an array).
+This can lead to integer overflows and thus incorrect results beyond precision.
+In many cases this may be silent, although NumPy usually gives warnings for the
+scalar operators.
+
+Similarliy, if the storage array is float32 a calculation may retain the lower
+float32 precision rather than use the default float64.
+
+Further issues can occure.  For example:
+
+* Floating point comparisons, especially equality, may change when mixing
+  precisions:
+  ```python3
+  np.float32(1/3) == 1/3  # was False, will be True.
+  ```
+* Certain operations are expected to start failing:
+  ```python3
+  np.array([1], np.uint8) * 1000
```

Comment:
Can you clarify, this fails because `1000` is not representable in `uint8`. But am I understanding correctly that `np.array([1], np.uint8) * 100` would silently overflow, because `100` is representable in `uint8`?

My main area of confusion from the set of examples here is which operations will silently overflow and which will error.

Reply:
Yes, that is what I have in mind currently.  I am not at all set on it though – it would actually be much simpler to not do that!
Basically, my rough plan would be that the above would fail, but `np.uint8(1000)` would not for now, so that you could still write `np.uint8(-1)` if you must for some reason – because... backward compatibility mostly.

That is painful, so a simpler middle ground could also be that both `np.uint8(1000)` and `np.uint8(1) + 1000` would give the same warning.

In general, NumPy does not give reliable warnings for integer overflows.  We only give warnings for scalar python operations (and even then randomly, because scalar operator calls are poorly organized).  So `np.uint8(10) + 250` will give a warning, but that example would not (it works with an array).

That is one argument for why things could be a smoother transition if we limit to Python operators:  For those (and scalars which are currently the place that upcasts a lot), we could pull off to ensure warnings are given.  (Super painful because it requires cleaning up the scalar mess, but a clear path.)

One thing that I am unsure about what JAX does, is that I don't really think we should tag on the "weak" property to the array as JAX does.  One, because I am worried it may be brittle/surprising to carry it around for long and second because unlike JAX our arrays are mutable, so there is an additional complexity.  (We can still tag on something as an implementation detail.)

Reply:
Thanks - regarding JAX & the weak type attribute: the reason we used that approach is that in JAX's JIT-compiled code, everything is represented by a JAX array (scalar inputs get automatically promoted). So to maintain the same behavior between JIT and non-JIT code paths, we need a way to mark a JAX array as weakly-typed. Since NumPy doesn't have that requirement, it seems that it would make sense to avoid it.

That said, one benefit of JAX's approach is that calling `jnp.asarray` on a function's inputs does not change the results even if the input is a scalar, because `asarray` maintains the weakness of the array.

On the other hand, allowing zero-dimensional arrays to be weak readily leads to cases where larger arrays can be weak as well (e.g. `jnp.broadcast_to(jnp.array(1.0), (100, 100))` returns a weakly-typed matrix) so weak types can no longer be thought of as strictly synonymous with scalars. That's a bit surprising, and is one reason we're currently actively discussing design considerations around the semantics of weakly-typed values.

Reply:
Ah, that is interesting (and I should mention the weak attribute in alternatives, even if I currently would need some convincing to consider it more seriously – and add 1-2 more examples here for clarity of the initial point, of course).

The broadcast example is nicely clear about how things may get confusing when dragging on the flag.  Long ago, I had early plans to support, in theory, even `[1, 2, 3]` to be considered "weak" – e.g. for ufuncs – but I had abandoned them: it seemed just to much and also way too many moving parts at the time.

In the end, it would be nice to have a clear concept of how far the "weakness" reaches.  I could half imagine an `np.asarray(x, allow_weak=True)` for library use, but I somewhat imagine libraries are better of handling it up-front with `np.result_type` (or by not calling `asarray` at all) – possibly `asarray_unless_literal(x)` if need be.

> That's a bit surprising, and is one reason we're currently actively discussing design considerations around the semantics of weakly-typed values.

It will be interesting to hear about that!  NumPy might end up being a bit more limited in options due to backcompat, but it would be nice if we can manage to align some concepts, i.e. at which step "weakness" is forgotten.